### PR TITLE
[distributed] implement all_gather for arbitrary python objects

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28811 [distributed] implement all_gather for arbitrary python objects**

Per https://github.com/pytorch/pytorch/issues/23232, we'd like to add support for collective communication functions working with Python objects. This PR implements `all_gather` for picklable python objects. Later diffs will implement support for the other primitives, but I wanted to get this out to agree on the high-level approach. This is implemented by pickling the objects into `ByteTensor`s, and then broadcasting them using `dist.all_gather`. We also store the size. This allows us to reconstruct the object.

Differential Revision: [D17889300](https://our.internmc.facebook.com/intern/diff/D17889300/)

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1122,6 +1144,37 @@ def all_gather_multigpu(output_tensor_lists,
         work.wait()
 
 
+def all_gather_object(object, group=group.WORLD, async_op=False):
+    """
+    Performs all_gather on a picklable object.
+    """
+    if _rank_not_in_group(group):
+        return
+    _check_default_pg()
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py]: Do we still need this if `group != group.WORLD`?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -1122,6 +1144,37 @@ def all_gather_multigpu(output_tensor_lists,
...
+    all_gather(object_size_list, size, group=group)
+
+    # create the list of output tensors needed for all_gather
+    max_object_size = max(object_size_list)
+    tensor_list = [torch.ByteTensor(size=(max_object_size,)) for _ in range(group_size)]
+
+    if size != max_object_size:
+        # all tensors input into gather should be of the same size, so pad the local tensor if it is not the max size.
+        padding = torch.ByteTensor(size=(max_object_size - size,))
+        byte_tensor = torch.cat((byte_tensor, padding), dim=0)
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py]: Does this mean we are creating 3 tensor storages here? original `byte_tensor`, `padding`, and the padded `byte_tensor`. I wonder if it would be faster if we directly allocate an empty tensor of the max size, and then copy the original `byte_tensor` to it.

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -1122,6 +1144,37 @@ def all_gather_multigpu(output_tensor_lists,
...
+
+    all_gather(tensor_list, byte_tensor, group=group, async_op=async_op)
+    # unpickle the tensors back into objects.
+    objects = [_tensor_to_object(tensor, object_size) for object_size, tensor in zip(object_size_list, tensor_list)]
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py]: What if this is an async_op? data in tensor_list might not be ready yet.

**Review by mrshenli — COMMENTED** [review]
This looks great to me. Thanks for working on this! 

The API of the async version is unclear yet. I am not sure if we need a new work/future whose `wait()` API returns a list of objects. Or, we could take an output list as arg and populate that when the work is done. The latter can be implemented in the existing `ProcessGroup::Work` in C++. But the former would need something different. 

cc @pietern

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/distributed_c10d.py]: It looks like we can first allocate a tensor of `max_size`, and copy the serialized object tensor into it. However, we do need to know `max_size` first, which requires creating the serialized object tensor and getting its size via `numel()`.

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (pietern) [reply torch/distributed/distributed_c10d.py]: Let's not support async operations for the "object collectives" to begin with.

**Review by pietern — COMMENTED** [review]
The tests should go in `test_distributed.py`, since it'll be a feature of `torch.distributed`.

I'd like these functions to go into a new file (e.g. `object.py` or something similar).

I took a stab at this in #22490 but never finished it...

**Comment by mrshenli** [comment]
Got another request for this feature: https://discuss.pytorch.org/t/how-to-gather-a-list-of-strings-from-different-rank-machines-in-ddp-mode-with-nccl-backend/86707

Shall we continue working on this PR?

**Comment by rohan-varma** [comment]
@mrshenli Thanks for the reminder! I placed this on the back burner but I should get some cycles soon to take another stab at it.

**Comment by sherylwang** [comment]
Hi, is there timeline for this pull request? Look forward to this feature!

**Comment by rohan-varma** [comment]
Closing this stale PR in favor of https://github.com/pytorch/pytorch/pull/42189
